---
title: "phyloseq analysis for blue-jam project: family level"
author: "jen nguyen"
date of last update: "10/25/2021"
output: html_document
---

Questions:
   
1. Which taxa consistently change in abundance between mucus and tissue samples? Mucus and lumen?
2. Are any taxa only present in tissue samples? Only mucus? Only lumen?



```{r initialize, include=FALSE}
library(ggplot2)
library(here)
library(phyloseq)
library(Biostrings)
library(tidyverse)
theme_set(theme_bw())
```

```{r read sample metadata into data frame}
meta <- read_csv(here("data","meta","sample_metadata.csv"),
                      col_types = cols(
                        Name = col_character(),
                        Extraction = col_date(format = ""),
                        Mouse = col_character(),
                        Replicate = col_character(),
                        Location = col_character()#,
                        #Pair = col_double()
                      ))
df <- meta %>% column_to_rownames("Name")
```

```{r add "pair" variable to dataframe}
Pair <- c(1,1,1,2,2,2,3,3,3,4,4,5,5,6,6,7,7,NA,NA,NA,NA,NA)
df <- mutate(df,Pair)
```

```{r load dada2 outputs, 1st sequencing run}
seqtab <- readRDS(here("output","dada2","seqtable.rds")) # raw sequence table
seqtab.nochim <- readRDS(here("output","dada2","seqtable_nochimeras.rds")) # sequence table w/o chimeras
taxa <- readRDS(here("output","dada2","taxonomic_assignments.rds")) # taxonomic assignments
```

```{r construct phyloseq object from dada2 output}
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(df), 
               tax_table(taxa))
```

```{r ASV short names}
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps
```

```{r melt ps object into dataframe}
df <- ps %>%
  tax_glom("Family") %>%
  psmelt %>%
  with_groups(Sample, mutate, Proportion = Abundance / sum(Abundance))
#  group_by(Sample) %>%
#  mutate(Proportion = Abundance / sum(Abundance)) %>%
#  ungroup
```

```{r 1.0 mucus vs tissue, all pairs}
df %>%
  filter(Location %in% c('Mucus', 'Tissue')) %>%
  ggplot(aes(Location, Proportion, color = Family)) +
  geom_point() +
  geom_line(aes(group = Pair)) +
  facet_wrap(~Family, scales = 'free_y') +
  theme(legend.position = 'none')
```

```{r 1.1 lumen vs mucus, all pairs}
df %>%
  filter(Location %in% c('Luminal', 'Mucus')) %>%
  ggplot(aes(Location, Proportion, color = Family)) +
  geom_point() +
  geom_line(aes(group = Pair)) +
  facet_wrap(~Family, scales = 'free_y') +
  theme(legend.position = 'none')
```

```{r 1.3 lumen vs mucus vs tissue}
df %>%
  filter(Location %in% c('Luminal', 'Mucus', 'Tissue')) %>%
  ggplot(aes(Location, Proportion, color = Family)) +
  geom_point() +
  geom_line(aes(group = Pair)) +
  facet_wrap(~Family, scales = 'free_y') +
  theme(legend.position = 'none')
```

```{r 1.4 isolate families with interesting signal}
df %>% 
  filter(Family %in% c('Acholeplasmataceae','Akkermansiaceae','Barnesiellaceae','Lachnospiraceae','Lactobacillaceae','Muribaculaceae')) %>%
  filter(Location %in% c('Luminal', 'Mucus', 'Tissue')) %>%
  ggplot(aes(Location, Proportion, color = Family)) +
  geom_point() +
  geom_line(aes(group = Pair)) +
  facet_wrap(~Family, scales = 'free_y') +
  theme(legend.position = 'none') +
  ggtitle("Selected families, 1st seq")
```




# basic overview of 1st dataset
```{r 0.1 bacterial families by location}
df %>%
  filter(Location %in% c('Luminal', 'Mucus', 'Tissue')) %>%
  ggplot(aes(x=Sample, y=Proportion, fill= Family)) + 
  geom_col() + facet_wrap(~Location, scales = 'free_x') #+
  #Notheme(legend.position = 'none')
```

```{r 0.2 bacterial families in controls}
df %>%
  filter(Sample %in% c('JN18', 'mock-','mock-2', 'NTC')) %>%
  ggplot(aes(x=Sample, y=Proportion, fill= Family)) + 
  geom_col() + facet_wrap(~Mouse, scales = 'free_x') +
  theme(legend.position = 'none')
```

# old stuff
```{r q2.1 all unique ASVs per mouse are present in all samples from that mouse}

# q2.1: per mouse, calculate mean and st dev of ASV presence across samples

mdf <- df %>%
  filter(Mouse == 3)
  
mdf = mdf %>% 
  with_groups(OTU, mutate, MeanProp = mean(Proportion)) 

mdf = mdf %>% 
  with_groups(OTU, mutate, SDProp = sd(Proportion)) 

mdf %>% 
  ggplot(aes(x=OTU, y=MeanProp, color = Family)) +
  geom_point() +
  geom_errorbar(aes(ymin=MeanProp-SDProp, ymax=MeanProp+SDProp), width=.2,
                 position=position_dodge(.9)) +
  theme(legend.position = 'none')

```
This method doesn't work!! Each ASV is present as a row, even when abundance is zero!


```{r q2.3 Xanthobacteraceae & Comamonadaceae read counts}

# q2.3: what are the read counts of specific taxa across locations? 

df %>% 
  filter(Family %in% c('Comamonadaceae', 'Xanthobacteraceae')) %>%
  ggplot(aes(x=Sample, y=Proportion, fill = Family)) + 
  geom_col() + 
  facet_grid(Family ~ ., scale = 'free_y') +
  theme(legend.position = 'none')

```
```{r q2.4 read counts for all families with summed proportions over 0.1}

# q2.4: what are the read counts of all abundant taxa across locations? 
#       "abundant" here is considered SumProp > 0.1
#       where SumProp is the sum of Proportions across all samples per ASV (OTU)


df %>% 
  filter(Mouse < 7) %>% 
  with_groups(OTU, mutate, SumProp = sum(Proportion)) %>% 
  filter(SumProp > 0.1) %>% 
  ggplot(aes(x=Sample, y=Proportion, fill = Family)) + 
  geom_col() + 
  facet_grid(Family ~ Mouse, scale = 'free') #+
  #theme(legend.position = 'none')
```
Families:
1. Acholeplasmataceae
2. Akkermansiaceae
3. Bacteroidaceae
4. Bifidobacteriaceae
5. Clostridiaceae
6. Enterobacteriaceae
7. Erysipelotrichaceae
8. Lachnospiraceae
9. Lactobacillaceae
10. Muribaculaceae
11. Oscillospiraceae
12. Peptococcaceae
13. Prevotellaceae
14. Rikenellaceae
15. Ruminococcaceae


```{r q2.4.1 families enriched in M + T}

df %>% 
  filter(Mouse < 7) %>% 
  with_groups(OTU, mutate, SumProp = sum(Proportion)) %>% 
  filter(SumProp > 0.1) %>% 
  filter(Family %in% c('Akkermansiaceae', 'Bacteroidaceae', 'Enterobacteriaceae','Lachnospiraceae','Prevotellaceae','Ruminococcaceae')) %>%
  ggplot(aes(x=Sample, y=Proportion, fill = Family)) + 
  geom_col() + 
  facet_grid(Family ~ Mouse, scale = 'free') +
  theme(legend.position = 'none')

```
```{r q2.4.2 families enriched in lumen}

df %>% 
  filter(Mouse < 7) %>% 
  with_groups(OTU, mutate, SumProp = sum(Proportion)) %>% 
  filter(SumProp > 0.1) %>% 
  filter(Family %in% c('Bifidobacteriaceae', 'Clostridiaceae', 'Lactobacillaceae')) %>%
  ggplot(aes(x=Sample, y=Proportion, fill = Family)) + 
  geom_col() + 
  facet_grid(Family ~ Mouse, scale = 'free') +
  theme(legend.position = 'none')
```

```{r q2.4.1 families of mixed location}

df %>% 
  filter(Mouse < 7) %>% 
  with_groups(OTU, mutate, SumProp = sum(Proportion)) %>% 
  filter(SumProp > 0.1) %>% 
  #filter(Family %in% c('Acholeplasmataceae', 'Erysipelotrichaceae','Muribaculaceae','Oscillospiraceae', 'Peptococcaceae','Rikenellaceae')) %>%
  filter(Family %in% c('Acholeplasmataceae', 'Muribaculaceae')) %>%
  ggplot(aes(x=Sample, y=Proportion, fill = Family)) + 
  geom_col() + 
  facet_grid(Family ~ Mouse, scale = 'free') +
  theme(legend.position = 'none')
```


# scratch work
```{r confirm that "abundance" in melted ps is post-filtering reads per sample}
ps.m1 = subset_samples(ps, Mouse == "1")
m1 <- psmelt(ps.m1) %>%
  ggplot(aes(x=Location, y=Abundance, fill=Phylum)) +
  geom_col()
print(m1)
```


```{r analysis1-1}
m1r <- ps.m1 %>%
  # transform abundance from read counts to rel abundance per sample
  tax_glom("Phylum") %>%
  transform_sample_counts(function(OTU) OTU/sum(OTU)) %>%
  psmelt %>%
  ggplot(aes(x=Location, y=Abundance, fill=Phylum)) + 
  geom_col() #+ facet_wrap(~Mouse)
print(m1r)
```

Questions inspired from the above plot:
Are the Cyanobacteria and Verricomicrobiota real? Or contamination?
Are Bacteroidota generally more abundant in crypts/tissue than Firmicutes? Or is this noise from low read count?



```{r analysis1.4 read counts across samples}
df %>%
  ggplot(aes(x=Sample, y=Abundance, fill= Family)) + 
  geom_col() + facet_wrap(~Mouse,scales = 'free_x') +
  theme(legend.position = 'none')
```

```{r analysis1.5 community proportions across samples}
df %>%
  ggplot(aes(x=Sample, y=Proportion, fill= Family)) + 
  geom_col() + facet_wrap(~Mouse,scales = 'free_x') +
  theme(legend.position = 'none')
```

```{r analysis2 luminal vs mucus, all 3 pairs}
df %>%
  filter(Location %in% c('Luminal', 'Mucus'), Replicate != 1) %>%
  ggplot(aes(Location, Proportion, color = Family)) +
  geom_point() +
  geom_line(aes(group = Mouse)) +
  facet_wrap(~Family, scales = 'free_y') +
  theme(legend.position = 'none')
```

```{r analysis3.0 mucus vs tissue, 6 pairs}
df %>%
  filter(Location %in% c('Mucus', 'Tissue'), Replicate != 1) %>%
  ggplot(aes(Location, Proportion, color = Family)) +
  geom_point() +
  geom_line(aes(group = Mouse)) +
  facet_wrap(~Family, scales = 'free_y') +
  theme(legend.position = 'none')
```
```{r analysis3.1 mucus vs. tissue, 3 gentle pairs}
df %>%
  filter(Location %in% c('Mucus', 'Tissue'), Extraction == "2021-07-23", Replicate != 1) %>%
  ggplot(aes(Location, Proportion, color = Family)) +
  geom_point() +
  geom_line(aes(group = Mouse)) +
  facet_wrap(~Family, scales = 'free_y') +
  theme(legend.position = 'none')
```
Increased:
  1. [1,2] Acholeplasmataceae
  2. [1,4] Akkermansiaceae
  3. [3,3] Erysipelotrichaeceae
  4. [4,4] Muribaculaceae

Decreased:
  1. [1,5] Anaerovoracaceae (though lowly abundant)

No change/not really there:
  1. [1,1] Coprostanoligenes
  2. [1,6] Bacillaceae
  3. [2,2] Bifidobacteriaceae
  4. [2,3] Butyricicoccaceae
  5. [2,4] Christensenellaceae
  6. [3,1] Eggerthellaceae
  7. [3,2] Enterobacteriaceae
  8. [3,4] Fusobacteriaceae
  9. [3,5] Lachnospiraceae
  10. [3,6] Lactobacillaceae
  11. [3,7] Methanobacteriaceae
  12. [4,1] Microbacteriaceae
  13. [4,2] Mitochondria
  14. [4,3] Monoglobaceae
  15. [5,1] Planococcaceae
  16. [5,3] Pseudomonadaceae
  17. [5,6] Sutterellaceae
  18. [5,7] UCG-010
  19. [6,2] Xanthobacteraceae

Present, but noisy:
  1. [1,3] Acidaminococcaceae
  2. [1,7] Bacteroidaceae
  3. [1,8] Barnesiellaceae
  4. [2,5] Clostridiaceae
  5. [2,6] Comanmonadaceae
  6. [2,7] Defluviitaleaceae
  7. [4,5] Oscillospiraceae
  8. [4,6] Peptococcaceae
  9. [4,7] Peptostreptococcaceae
  10. [5,2] Prevotellaceae
  11. [5,4] Rikenellaceae
  12. [5,5] Ruminococcaceae
  13. [6,1] Veillonellaceae

```{r analysis3.2 mucus vs tissue, 3 hard pairs}
df %>%
  filter(Location %in% c('Mucus', 'Tissue'), Extraction == "2021-07-30") %>%
  ggplot(aes(Location, Proportion, color = Family)) +
  geom_point() +
  geom_line(aes(group = Mouse)) +
  facet_wrap(~Family, scales = 'free_y') +
  theme(legend.position = 'none')
```
Do trends between gentle or harder scraping preps change? What could lead to change?
1. Decrease suggests more mucus (or tissue) was removed and the bacteria along with it
2. Increase suggests reduction in mucus (or tissue) made remaining bugs more abundant
3. More noise suggests additional changes in direction that were not consistent with previous trend.

Example: [3,3] Erysipelotrichaeceae was increased in tissue from gentle scrapings but mixed (increase, decrease and no change) in hard scraping samples.
It is a quite abundant family, making up ~10% of ASVs in one particular tissue sample. 5-15% of ASVs in the three luminal samples.

- could looking beyond family clear up some of the noise?


Increased:
  1. [1,2] Acholeplasmataceae
  2. [1,4] Akkermansiaceae
  3.
  4. [4,4] Muribaculaceae
  
  changed between analysis 3.1 and 3.2
  5. [5,4] Rikenellaceae
  6. [5,6] Sutterellaceae

Decreased:
  1. 
  
  changed between analysis 3.1 and 3.2
  2. [5,2] Prevotellaceae

No change/not really there:
  1. [1,1] Coprostanoligenes
  2. [1,6] Bacillaceae
  3. [2,2] Bifidobacteriaceae
  4. [2,3] Butyricicoccaceae
  5. [2,4] Christensenellaceae
  6. [3,1] Eggerthellaceae
  7. [3,2] Enterobacteriaceae
  8. [3,4] Fusobacteriaceae
  9. [3,5] Lachnospiraceae
  10. [3,6] Lactobacillaceae
  11. [3,7] Methanobacteriaceae
  12. [4,1] Microbacteriaceae
  13. [4,2] Mitochondria
  14. [4,3] Monoglobaceae
  15. [5,1] Planococcaceae
  16. [5,3] Pseudomonadaceae
  17. 
  18. [5,7] UCG-010
  19. [6,2] Xanthobacteraceae

Present, but noisy:
  1. [1,3] Acidaminococcaceae
  2. [1,7] Bacteroidaceae
  3. [1,8] Barnesiellaceae
  4. [2,5] Clostridiaceae
  5. [2,6] Comanmonadaceae
  6. [2,7] Defluviitaleaceae
  7. [4,5] Oscillospiraceae
  8. [4,6] Peptococcaceae
  9. [4,7] Peptostreptococcaceae
  10. 
  11. 
  12. [5,5] Ruminococcaceae
  13. [6,1] Veillonellaceae
  
  changed between 3.1 and 3.2
  14. [3,3] Erysipelotrichaeceae
  15. [1,5] Anaerovoracaceae



```{r alpha diversity}
plot_richness(ps, x="sample.names", measures=c("Shannon", "Simpson"), color="Location")
```
Harder scrapings tend to have higher diversity in tissue than in mucus.

RE: the message from phyloseq when running this data:
"the data you have provided does not have any singletons. This is highly suspicious. Results of richness
estimates (for example) are probably unreliable, or wrong, if you have already trimmed low-abundance taxa from the data."

Outputs from DADA2 are unlikely to have singletons because single reads are often considered as error.
Especially single reads that are very similar to another more common read, DADA2 often interprets this as sequencing error.
Were we to expect some very lowly abundant organisms in some samples, we could consider a technique called "pseudopooling", which considers whether the single reads are found in different samples on the same plate. This suggests they are true sequence variants!

Overall, considering singletons only matters if we want to interpret them as reads.


```{r Bray Curtis}
# Transform data to proportions as appropriate for Bray-Curtis distances
ps.prop <- transform_sample_counts(ps, function(otu) otu/sum(otu))
ord.nmds.bray <- ordinate(ps.prop, method="NMDS", distance="bray")
```

```{r}
# Bray-Curtis ordination
plot_ordination(ps.prop, ord.nmds.bray, color="Location", title="Bray NMDS")
```

Luminal samples cluster together, and controls (mock, extraction and water) are grouped together away from most mouse samples.
Tissue and mucus samples are not easily distinguished however.
Which points belong to which pairs?
Does one method of scraping group in pairs and one method seem affect their likeness?






```{r confirm that "abundance" in melted ps is post-filtering reads per sample}
ps.m1 = subset_samples(ps, Mouse == "1")
mouse1 = psmelt(ps.m1)
colnames(mouse1)
m1 = ggplot(mouse1, aes(x=Location, y=Abundance, fill=Phylum))
m1 = m1 + geom_bar(color="black", stat="identity", position="stack") #+ facet_wrap(~Mouse)
print(m1)
```

```{r analysis1}
# transform abundance from read counts to rel abundundance per sample
ps.m1.rel <- transform_sample_counts(ps.m1, function(OTU) OTU/sum(OTU)) 
mouse1.rel = psmelt(ps.m1.rel)
colnames(mouse1.rel)
m1r = ggplot(mouse1.rel, aes(x=Location, y=Abundance, fill=Phylum))
m1r = m1r + geom_bar(color="black", stat="identity", position="stack") #+ facet_wrap(~Mouse)
print(m1r)
```

```{r}
# Bar plot, top 20 most abundant families
top20 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:20]
ps.top20 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top20 <- prune_taxa(top20, ps.top20)
plot_bar(ps.top20, x="Mouse", fill="Family") + facet_wrap(~Location, scales="free_x")
```

Plotting the top 20 families only gets us <75% of reads in some samples, <25% of reads than in others. Try plotting more families, say 40. See below!


```{r}
# Bar plot, top 40 most abundant families
top40 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:40]
ps.top40 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top40 <- prune_taxa(top40, ps.top40)
plot_bar(ps.top40, x="Mouse", fill="Family") + facet_wrap(~Location, scales="free_x")
```
Hm, after including most abundant 40 families, now we see abundances of over 1... 
Something is not right. What are we actually plotting?




```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.


```{r}
sessionInfo()
```

